### PR TITLE
Implement tcp transport.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/tcp.toit [hunk 1]
```diff
@@ -5,29 +5,118 @@
...
+  reconnect -> none:
+    throw "UNSUPPORTED"
+
+class ReconnectingTransport_ extends TcpTransport:
+  // Reconnection information.
+  interface_ /tcp.Interface
+  host_      /string
+  port_      /int
+
+  reconnecting_mutex /monitor.Mutex := monitor.Mutex
```
Comment (kasperl) [inline src/tcp.toit:66]: It feels odd that this is public.

File: src/tcp.toit [hunk 2]
```diff
@@ -5,29 +5,118 @@
...
+
+  write bytes/ByteArray -> int:
+    return socket_.write bytes
+
+  read -> ByteArray?:
+    return socket_.read
+
+  reconnect:
+    old_socket := socket_
+    reconnecting_mutex = monitor.Mutex
```
Comment (kasperl) [inline src/tcp.toit:82]: If you always create a new mutex for every reconnect attempt, then what is the point?

File: src/tcp.toit [hunk 3]
```diff
@@ -5,29 +5,118 @@
...
+    reconnecting_mutex.do:
+      if not identical old_socket socket_: return
+      if old_socket: old_socket.close
+
+      socket := new_connection_
+      // Send messages immediately.
+      socket.set_no_delay true
```
Comment (kasperl) [inline src/tcp.toit:89]: This will change to `socket.no_delay = true` pretty soon. Maybe you could do it already now?

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/tcp.toit [hunk 4]
```diff
@@ -5,29 +5,118 @@
...
-      with_timeout timeout:
-        return Packet.deserialize reader_
-    return null
+class TcpTransport implements Transport BrokerTransport:
+  socket_ /tcp.Socket? := null
+
+  constructor socket/tcp.Socket:
+    socket_ = socket
+
+  constructor interface/tcp.Interface --host/string --port/int=1883:
```
[the remark this answers is not in the document]
Comment (floitsch) [reply src/tcp.toit:27]: done.

**Pull request merged**